### ISSUE TITLE
[Feature request] Sorted symfile

### ISSUE THREAD

**Issue by aaaaaa123456789** [rednex/rgbds#215]
As of right now, symfiles contain symbols in the order the linker encounters them. That results in unsorted lists that are unsuitable for human consumption.

It would be very useful if symfiles were sorted by section type, bank, address and symbol name, in that order. (Symbol name is used to break ties between identical addresses in order to ensure that symbols are always outputted in the same order.) For instance, the following source file:

```
SECTION "rom0", ROM0[0]
A:: ds $1000
B:: ds $2000
C:: ds $1000
D::

SECTION "rom1", ROMX[$4000], BANK[1]
E:: ds $2800
F:: ds $1800
G::

SECTION "rom2", ROMX[$4000], BANK[2]
H:: ds $4000
I::

SECTION "vram0", VRAM[$8000], BANK[0]
J:: ds $1000
K: ds $1000
L::

SECTION "vram1", VRAM[$8000], BANK[1]
M: ds $2000
N::

SECTION "sram0", SRAM[$a000], BANK[0]
O:: ds $1800
P:: ds $800
Q::

SECTION "sram1", SRAM[$a000], BANK[1]
R:: ds $1000
S:: ds $1000
T::

SECTION "wram0", WRAM0[$c000]
U:: ds $1000
V::

SECTION "wram1", WRAMX[$d000], BANK[1]
W:: ds $800
X:: ds $800
Y::

SECTION "wram2", WRAMX[$d000], BANK[2]
Z::
AA:: ds $1000
AB::

SECTION "hram", HRAM[$ff80]
AC:: ds $40
AD:: ds $3f
AE::
```

...should always generate the following symfile, regardless of section reordering:

```
00:0000 A
00:1000 B
00:3000 C
00:4000 D
01:4000 E
01:6800 F
01:8000 G
02:4000 H
02:8000 I
00:8000 J
00:9000 K
00:A000 L
01:8000 M
01:A000 N
00:A000 O
00:B800 P
00:C000 Q
01:A000 R
01:B000 S
01:C000 T
00:C000 U
00:D000 V
01:D000 W
01:D800 X
01:E000 Y
02:D000 AA
02:D000 Z
02:E000 AB
00:FF80 AC
00:FFC0 AD
00:FFFF AE
```

(Note that section type as a first sorting criterion is _not_ implied by bank and address. This is why `01:A000 N` appears much earlier than `01:A000 R` — the former is VRAM, the latter is SRAM.)

**Comment by mid-kid** [comment]
Note that the symfile _doesn't_ list any unreferenced labels, or labels only referenced through `jr`, which is probably something we want.

**Comment by AntonioND** [comment]
You can use `-E` to export all labels, have you tried that?

**Comment by mid-kid** [comment]
Due to conflicts in local label names across different files, that isn't a viable option as it breaks linking.

**Comment by AntonioND** [comment]
Then there's no easy way to fix that problem. It doesn't make sense to only export only some local labels, you should change the names. A proper fix is trickier, as some projects rely on that behaviour to link, as far as I know.

**Comment by mid-kid** [comment]
Oh well. Thanks for explaining!

**Comment by aaaaaa123456789** [comment]
Is there any command-line flag that sorts them, as asked originally?

**Comment by AntonioND** [comment]
No, and I'm working on #203 right now.  As a workaround you can just use `sort file.sym > file_sorted.sym` or something like that.

**Comment by AntonioND** [comment]
While at it, I guess that the mapfile could finally be printed correctly and ignore all banks that aren't used. Right now, for example, all 16 SRAM banks are printed, even though they don't have a reason to.

**Comment by ISSOtm** [comment]
Fixed by #434.
